### PR TITLE
audio-only: implement initial "low bandwidth mode"

### PR DESCRIPTION
**Pull request by saghul**
It's an evolution of audio-only mode, where we also allow for receiving a remote
screen-share.

Diving deeper: this basically sets last N to 1 or 0 depending on the
availability of a screen-share.

### CODE REVIEW

**Comment by saghul** [comment]
The looks:

Screen Shot 2019-08-06 at 15 08 23
Screen Shot 2019-08-06 at 15 08 01
Screen Shot 2019-08-06 at 15 02 11

File: modules/UI/videolayout/SmallVideo.js [hunk 1]
```diff
@@ -557,24 +557,19 @@ SmallVideo.prototype._isHovered = function() {
...
- *
- * @param show whether we should show the avatar or not
- * video because there is no dominant speaker and no focused speaker
  */
 SmallVideo.prototype.updateView = function() {
     if (this.disableUpdateView) {
         return;
     }
 
-    if (!this.hasAvatar) {
```
Comment (zbettenbuk) [inline modules/UI/videolayout/SmallVideo.js]: Why is this change? This doesn't seem to be related to LB mode

File: react/features/base/lastn/middleware.js [hunk 2]
```diff
@@ -52,12 +58,27 @@ function _updateLastN({ getState }) {
...
+    if (appState !== 'active') {
         lastN = 0;
+    } else if (audioOnly) {
+        const { screenShares, tileViewEnabled } = state['features/video-layout'];
+        const largeVideoParticipantId = state['features/large-video'].participantId;
+        const largeVideoParticipant
+            = largeVideoParticipantId ? getParticipantById(state, largeVideoParticipantId) : undefined;
+
+        if (!tileViewEnabled && largeVideoParticipant && !largeVideoParticipant.local) {
+            lastN = (screenShares || []).includes(largeVideoParticipantId) ? 1 : 0;
```
Comment (zbettenbuk) [inline react/features/base/lastn/middleware.js]: How do you make sure we don't end up rendering a non-screenshare video?

**Comment by saghul** [comment]
Screen Shot 2019-08-06 at 15 02 11

File: react/features/base/participants/functions.js [hunk 3]
```diff
@@ -299,44 +299,50 @@ export function isLocalParticipantModerator(
...
+        return true;
+    }
+
+    /* Last, check if the participant is sharing their screen and they are on stage. */
+    const screenShares = state['features/video-layout'].screenShares || [];
+    const largeVideoParticipantId = state['features/large-video'].participantId;
+    const participantIsInLargeVideoWithScreen
+        = participant.id === largeVideoParticipantId && screenShares.includes(participant.id);
+
+    return participantIsInLargeVideoWithScreen;
```
Comment (zbettenbuk) [inline react/features/base/participants/functions.js]: Are you sure it will end up with the same logic (plus screenshare)? That ```shouldRenderVideoTrack``` function seemed logical to me. How do you substitute that here?

File: modules/UI/videolayout/SmallVideo.js (hunk 1 above)
Comment (saghul) [reply modules/UI/videolayout/SmallVideo.js]: It's in a separate commit. The screen-share icon looks tiny otherwise.

File: react/features/base/lastn/middleware.js (hunk 2 above)
Comment (saghul) [reply react/features/base/lastn/middleware.js]: Exactly there. If the participant in large video is not part of the screen sharing set, set last N to 0 (no video) otherwise set to 1. When set to 1, the selected participant (aka, the one in the large view) will be displayed.

File: react/features/base/participants/functions.js (hunk 3 above)
Comment (saghul) [reply react/features/base/participants/functions.js]: Good point. On mobile (where this function is used). That is just `videoTrack && !videoTrack.muted`. On web it gets trickier because it waits for a frame and remembers that and so on.

But if you think it's clearer if I bring it back, I will.

Comment (saghul) [reply react/features/base/participants/functions.js]: Done.

**Comment by paweldomas** [comment]
Does this logic also work on web ? What happens if you start screensharing while in the low bandwidth mode ?

**Comment by saghul** [comment]
Yes, this works also on web. If you are in low bw mode and enable screen sharing it behaves the same as if you unmute video: you exit the mode.

We may want to refine some of the behavior, this is just step one, if you will.

Thanks a lot for the review!

File: modules/UI/videolayout/SmallVideo.js (hunk 1 above)
Comment (hristoterezov) [inline modules/UI/videolayout/SmallVideo.js]: @saghul It seems `this.hasAvatar` is not used anymore but has a few setters here and there. Am I correct?

Comment (saghul) [reply modules/UI/videolayout/SmallVideo.js]: It’s possible this was the last use. Sorry I didn’t notice!
